firehose/firehose_kinesis.go

```diff
@@ -0,0 +1,125 @@
+package firehose
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/Clever/heka-clever-plugins/aws"
+	"github.com/Clever/heka-clever-plugins/batcher"
+)
+
+type FirehoseWriter struct {
+	conf     *FirehoseWriterConfig
+	batchers map[string]batcher.Batcher
+
+	mockEndpoint string
+
+	recvRecordCount    int64
+	sentRecordCount    int64
+	droppedRecordCount int64
+}
+
+type FirehoseWriterConfig struct {
+	// The value of this field is used as the firehose stream name
+	StreamName string
+	// AWS region the stream lives in
+	Region string `toml:"region"`
+	// Interval at which accumulated messages should be bulk put to
+	// firehose, in milliseconds (default 1000, i.e. 1 second).
+	FlushInterval uint32 `toml:"flush_interval"`
+	// Number of messages that triggers a push to firehose
+	// (default to 1, maximum is 500)
+	FlushCount int `toml:"flush_count"`
+	// Size of batch that triggers a push to firehose
+	// (default to 1024 * 1024 (1mb))
+	FlushSize int `toml:"flush_size"`
+}
+
+func NewFirehoseWriter(config FirehoseWriterConfig, mockEndpoint string) (*FirehoseWriter, error) {
+	if config.FlushCount > 500 || config.FlushCount < 1 {
+		return nil, fmt.Errorf("FlushCount must be between 1 and 500 messages")
+	}
+	return &FirehoseWriter{
+		conf:         &config,
+		mockEndpoint: mockEndpoint,
+		batchers:     map[string]batcher.Batcher{},
+	}, nil
+}
+
+func (f *FirehoseWriter) ProcessMessage(msg string) error {
+	atomic.AddInt64(&f.recvRecordCount, 1)
+
+	fields := convertToFields(msg)
+
+	record, err := json.Marshal(fields)
+	if err != nil {
+		atomic.AddInt64(&f.droppedRecordCount, 1)
+		return err
+	}
+
+	batch, ok := f.batchers[f.conf.StreamName]
+	if !ok {
+		sync := f.createBatcherSync(f.conf.StreamName)
+		batch = batcher.New(sync)
+		f.batchers[f.conf.StreamName] = batch
+	}
+	batch.Send(record)
+
+	return nil
+}
+
+func convertToFields(msg string) map[string]interface{} {
```

Comment:
removed convertToFields for now. will revisit when we add more decoding logic 8bef26211cce095b7b7d3bb5c6b8f40f46bfecfc